dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,407 @@
+guarantees assured by [Formal proofs](https://sel4.systems/Info/FAQ/proof.pml),
+however these proofs are still incomplete for x86, see
+[Supported Platforms](https://docs.sel4.systems/Hardware) for an up-to-date
+verification status. seL4 due to its microkernel nature provides higher
+isolation, breach in one of the components (like USB driver, network stack)
+wouldn't compromise entire OS contrary to monolithic kernels.
+
+We have opened an [issue](https://github.com/seL4/seL4/issues/832) about seL4
+usage as secure bootloader.
+
+#### Running in DLME
+
+It should be runnable, however due to microkernel nature all programs run in
+unprivileged mode which complicates booting of the target OS. seL4 would have to
+gain kexec-like abilities.
+
+#### Fobnail integration
+
+seL4 provides virtually no drivers, except a few drivers listed here.
+According to the page linked above, `libusbdrivers` is inactive and lacks XHCI
+support. seL4 has basic support for an old version of musl C library (v1.1.16).
+
+Using seL4 would require a significant amount of work:
+- Extending USB drivers
+- Implementing USB EEM driver
+- Implementing TPM driver
+- Porting [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+- seL4 runs all processes in unprivileged mode, so the kernel itself would have
+  to be modified to allow booting of another kernel
+
+[Genode](https://github.com/genodelabs/genode) is an framework for creating
+custom, specialized OSes. It provides USB drivers (including `usb-net` driver
+from which should handle USB EEM) and a network stack. Its support for seL4 used
+to be incomplete, and many components were broken. However, that might have
+changed, and Genode may be an easier way to get software running on seL4. We
+have opened an [issue](https://github.com/genodelabs/genode/issues/4480) here
+too.
+
+### Linux
+
+In the case of Linux, a minimal distribution will be prepared that meets the
+requirements of the project. [Yocto Project](https://www.yoctoproject.org/) will
+be used for this, because it gives a lot of freedom in manipulating the elements
+that make up the target system.
+
+#### Running in DLME
+
+Ensuring that the operating system works in DLME can be achieved by using the
+[TrenchBoot](https://trenchboot.org/) project. For this purpose, a Yocto
+[meta-fobnail](https://github.com/fobnail/meta-fobnail) layer has been created
+that integrates the necessary elements. The effects of the tests are presented
+in a [separate document](./running-os-in-dlme.md).
+
+#### Fobnail integration
+
+At this point, the [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+application is strongly dependent on Linux. Mainly due to the fact that its
+development took place on this operating system. Therefore, it is important that
+the created minimal OS also has an integrated attester application. The main
+dependencies of its operation are drivers for USB, USB EEM and TPM, but running
```

Comment:
It would be substantially simpler to use USB directly rather than tunneling TCP/IP over it.

Reply:
> `s/TCP/UDP/;s/TLS/DTLS/`

Code that relies on UDP to provide reliable, in-order delivery (as yours does) is already non-portable, and in fact is likely to be broken as drops can still take place in e.g. the network stack’s internal buffers.  If you are going to go with this approach I suggest using TLS over USB, which is perfectly valid as TLS is transport-agnostic.

Reply:
@DemiMarie We have some basic reliability layer which should handle lost packets and out of order delivery, however it hasn't been tested much.

Reply:
I don't think we have more to add here. @DemiMarie can we resolve this conversation? I am asking because it was started by you.

Reply:
I appreciate that, thank you.